### PR TITLE
Let SshKeyGenFragment use the Android Keystore

### PR DESCRIPTION
**Pull request by fmeum**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [x] Enhancement
- [ ] Refactoring

## :scroll: Description

**Note: This is ready for review, but requires a patched SSHJ dependency since version 0.30.0 hasn't been released yet. Once it is and the review is done, I will remove the `.jar` and depend on the Maven repository.**

Implement an SSH auth method backed by an Android Keystore key and make `SshKeyGenFragment` generate such a key. 

In detail, this means the following:

- SSH keys generated within the app are no longer kept in a simple file optionally encrypted with a passphrase, but rather in the Android Keystore optionally protected with the user's screen lock credential. On modern devices with a TEE this means that the private key will not be leaked even if the device is rooted or compromised remotely.
- SSH key files with passphrases remain supported for users who prefer them, they just can't be generated from within the app anymore.
- The offered key types are changed from RSA-2048 and RSA-4096 (default) to RSA-2048, RSA-3072 and P-384 (default) for the following reasons:
    - RSA-2048 is by far the most widely supported key type and reasonably fast. Although it is no longer recommended to be used for security needs that extend into the 2030s, this does not matter too much in the context of SSH authentication. It's a keep, but should not be the default.
    - RSA-3072 will remain secure for at least a decade or two and can't possibly be backdoored by the NSA, which likely appeals to a good part of our userbase.
    - RSA-4096, the previous default, is unreasonably slow. I am open to be convinced of its usefulness, but so far fail to see why this algorithm would ever be a good fit for authentication (I could see this being useful for long-term encryption though).
    - P-384 is both very fast and recommended for long-term security, which makes it a great default option. Of course not everybody will be happy with this (or any other NIST-supplied) curve, but for SSH authentication against a pass git repository, I don't see any actual issues.
- Small changes were made to `BiometricAuthenticator` to fix one of two memory leaks.
- I needed to duplicate some of SSHJ's code, but luckily its Apache 2.0 license is compatible with the GPL 3.

Note: SSHJ so far doesn't support SHA-256/512 signatures with `ssh-rsa` key types. This will probably be fixed upstream at some point, which is why we allow these signature algorithms to be used with our Keystore keys. Again, in the context of SSH authentication, relying on SHA1 is not a major issue.

## :bulb: Motivation and Context

We were not making use of the extensive Android capabilities around cryptography and had to rely on Jsch to generate SSH keys for us.

## :green_heart: How did you test it?
I generated and used both RSA and EC keys, please do the same.

## :pencil: Checklist

- [x] I formatted the code with the IDE's reformat action (Ctrl + Shift + L/Cmd + Shift + L)
- [x] I reviewed submitted code

## :camera_flash: Screenshots / GIFs
<details>
<summary>
Screenshots
</summary>

</details>

### CODE REVIEW

**Comment by FabianHenneke** [comment]
I should probably that the reason `ssh-ed25519` is not offered is purely because it is not yet supported by the Android Keystore.

**Review by msfjarvis — DISMISSED** [review]
LGTM. Needs a release notes entry.

**Review by msfjarvis — DISMISSED** [review]

**Comment by FabianHenneke** [comment]
> Release builds crash for me with `java.lang.NoClassDefFoundError: Failed resolution of: Lnet/i2p/crypto/eddsa/spec/EdDSANamedCurveTable;`, adding a keep rule to `proguard-rules.pro` didn't have any effect.
> 
> ```diff
> diff --git app/proguard-rules.pro app/proguard-rules.pro
> index 474b45d0065d..b6958c965e0f 100644
> --- app/proguard-rules.pro
> +++ app/proguard-rules.pro
> @@ -27,3 +27,4 @@
>  -keep class org.bouncycastle.jcajce.provider.** { *; }
>  -keep class org.bouncycastle.jce.provider.** { *; }
>  -keep class !org.bouncycastle.jce.provider.X509LDAPCertStoreSpi { *; }
> +-keep class net.i2p.crypto.eddsa.spec.EdDSANamedCurveTable { *; }
> ```

That seems to be an issue with the filesDir style of dependency declaration. I can only fix that by manually adding the dependencies of SSHJ to our build.gradle. But of course this will not be necessary once 0.30.0 is released.

**Comment by msfjarvis** [comment]
> > Release builds crash for me with `java.lang.NoClassDefFoundError: Failed resolution of: Lnet/i2p/crypto/eddsa/spec/EdDSANamedCurveTable;`, adding a keep rule to `proguard-rules.pro` didn't have any effect.
> > 
> >       
> >         
> >       
> >     ```diff
> > diff --git app/proguard-rules.pro app/proguard-rules.pro
> > index 474b45d0065d..b6958c965e0f 100644
> > --- app/proguard-rules.pro
> > +++ app/proguard-rules.pro
> > @@ -27,3 +27,4 @@
> >  -keep class org.bouncycastle.jcajce.provider.** { *; }
> >  -keep class org.bouncycastle.jce.provider.** { *; }
> >  -keep class !org.bouncycastle.jce.provider.X509LDAPCertStoreSpi { *; }
> > +-keep class net.i2p.crypto.eddsa.spec.EdDSANamedCurveTable { *; }
> > ```
> 
> That seems to be an issue with the filesDir style of dependency declaration. I can only fix that by manually adding the dependencies of SSHJ to our build.gradle. But of course this will not be necessary once 0.30.0 is released.

Can probably just apply the [shadow](https://github.com/johnrengelman/shadow) plugin and build a 'fat JAR' in the mean time.

**Comment by FabianHenneke** [comment]
I just looked into how we do password authentication via SSH, but am confused: Is this function https://github.com/android-password-store/Android-Password-Store/blob/a8f6669e35e8eb841c504269645b03dd4dce2494/app/src/main/java/com/zeapo/pwdstore/git/GitOperation.kt#L55 dealing with password authentication both via SSH and via HTTPS? It relies on an `SshSessionFactory`, but the code that calls it supplies the `https_password` to it. I don't want to break HTTPS when I switch the backend for SSH password authentication over to SSHJ.

**Comment by msfjarvis** [comment]
> I just looked into how we do password authentication via SSH, but am confused: Is this function
> 
> https://github.com/android-password-store/Android-Password-Store/blob/a8f6669e35e8eb841c504269645b03dd4dce2494/app/src/main/java/com/zeapo/pwdstore/git/GitOperation.kt#L55
> 
> dealing with password authentication both via SSH and via HTTPS? It relies on an `SshSessionFactory`, but the code that calls it supplies the `https_password` to it. I don't want to break HTTPS when I switch the backend for SSH password authentication over to SSHJ.

It's using an `SshSessionFactory` because it looks like HTTPS was tacked on later and this was just never fixed. It should not need that factory to work.

**Comment by msfjarvis** [comment]
I can't test the ED25519 key I generated on my PC with builds off this branch, what am I missing? The crash is caused by `java.lang.NoClassDefFoundError: Failed resolution of: Lnet/i2p/crypto/eddsa/spec/EdDSANamedCurveTable;` and reproducible on both debug and release binaries.

**Comment by FabianHenneke** [comment]
> I can't test the ED25519 key I generated on my PC with builds off this branch, what am I missing? The crash is caused by `java.lang.NoClassDefFoundError: Failed resolution of: Lnet/i2p/crypto/eddsa/spec/EdDSANamedCurveTable;` and reproducible on both debug and release binaries.

I think that is because of the way I build the SSHJ jar. Could you try adding the dependency (https://github.com/hierynomus/sshj/blob/master/build.gradle#L52) manually?

**Comment by msfjarvis** [comment]
> > I can't test the ED25519 key I generated on my PC with builds off this branch, what am I missing? The crash is caused by `java.lang.NoClassDefFoundError: Failed resolution of: Lnet/i2p/crypto/eddsa/spec/EdDSANamedCurveTable;` and reproducible on both debug and release binaries.
> 
> I think that is because of the way I build the SSHJ jar. Could you try adding the dependency ([hierynomus/sshj:build.gradle@`master`#L52](https://github.com/hierynomus/sshj/blob/master/build.gradle#L52)) manually?

On it

**Comment by msfjarvis** [comment]
Works on release builds with the explicit dependency.

**Comment by msfjarvis** [comment]
JitPack doesn't seem to have Java 9 runners which SSHJ mandates now so we can't rely on them to be able to eliminate this JAR either. Gonna have to wait it out for an SSHJ release...

**Comment by FabianHenneke** [comment]
I'm positive we'll get a new release soon, the current maintainer seems to be quite responsive.

**Comment by msfjarvis** [comment]
@FabianHenneke this is gonna need a rebase soon-ish, I wanted to test drive this a bit more to confirm that the opening the biometric dialog is not error-prone.

**Comment by FabianHenneke** [comment]
I rebased and along the way removed some unnecessary `getString` calls.

**Comment by msfjarvis** [comment]
Biometric dialog seems to open correctly on Android R as well as Samsung's OneUI 2.0 based on Android 10. Should be good to ship once we have an SSHJ release.

Will we need an FAQ page?

**Comment by FabianHenneke** [comment]
> Will we need an FAQ page?

The answer is probably yes. The UI already referenced an FAQ page when I started working on it. Was that still up to date?

**Comment by msfjarvis** [comment]
> > Will we need an FAQ page?
> 
> The answer is probably yes. The UI already referenced an FAQ page when I started working on it. Was that still up to date?

Nope, the data in there was very much antiquated and probably also deleted when I redid the wiki a few months ago.

**Comment by FabianHenneke** [comment]
On second thought: We could also remove the reference to the FAQ entirely. The error messages reported by SSHJ are very similar to the ones SSH would show (host key verification failed, authentication methods exhausted,...) and weird ones from JGit are unexplainable anyway.

However, I will update https://github.com/android-password-store/Android-Password-Store/wiki/Generate-SSH-Key and add a note about the implications of relying on the Keystore.

**Comment by msfjarvis** [comment]
> On second thought: We could also remove the reference to the FAQ entirely. The error messages reported by SSHJ are very similar to the ones SSH would show (host key verification failed, authentication methods exhausted,...) and weird ones from JGit are unexplainable anyway.

Yeah that sounds good.

> However, I will update https://github.com/android-password-store/Android-Password-Store/wiki/Generate-SSH-Key and add a note about the implications of relying on the Keystore.

:+1:

**Comment by FabianHenneke** [comment]
I refactored the key generation and added support for P-256 and the Pixel's StrongBox.

**Comment by msfjarvis** [comment]
I have some UX concerns around the dialog for the private key. Material strongly discourages the use of a neutral button, so we should ditch ok and replace it with copy. Secondly, we should probably not dismiss the calling activity, or at least not when we're viewing the key in settings

**Comment by FabianHenneke** [comment]
I hope I carried out what you suggested.

**Comment by msfjarvis** [comment]
Welp forgot to mention this: a Toast saying "copied to clipboard" and then dismissing the dialog would also go a long way. Users I dogfooded the builds on pointed out the lack of visible feedback indicating that copying had succeeded.

**Comment by FabianHenneke** [comment]
I added the toast.

**Comment by msfjarvis** [comment]
All good to go now, thanks!

File: app/src/main/java/com/zeapo/pwdstore/git/config/SshjConfig.kt
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/git/config/SshjConfig.kt]: I got rid of plenty of algorithms here compared to https://github.com/hierynomus/sshj/blob/master/src/main/java/net/schmizz/sshj/DefaultConfig.java to keep us in line with the [Mozilla Modern SSH config](https://infosec.mozilla.org/guidelines/openssh#modern). SSHJ does not have AES-GCM support (most likely a good thing given that it would have to be implemented in Java), but hopefully adopts ChaCha20-Poly1305 at some point.

Comment (msfjarvis) [reply app/src/main/java/com/zeapo/pwdstore/git/config/SshjConfig.kt]: I looked at the Mozilla config and I think it's a pretty sane collection, we should be fine. Would certainly love to have ChaCha20.

Comment (fmeum) [reply app/src/main/java/com/zeapo/pwdstore/git/config/SshjConfig.kt]: Unfortunately all the ciphers currently supported by SSHJ are unauthenticated. Getting ChaCha20-Poly1305 support in there would require some structural changes and thus support by the current maintainer. We should definitely revisit that point in the future.

**Comment by FabianHenneke** [comment]
I merged and confirmed that this works with RSA keys. Something in SSHJ seems to have broken ecdsa keys, I will look into that later, but you should be able to test with RSA keys.

I also found an issue we could and should fix ourselves: Our BiometricPrompt currently throws an exception when the fingerprint verification fails once.

**Comment by msfjarvis** [comment]
> I merged and confirmed that this works with RSA keys. Something in SSHJ seems to have broken ecdsa keys, I will look into that later, but you should be able to test with RSA keys.

Alright.

> I also found an issue we could and should fix ourselves: Our BiometricPrompt currently throws an exception when the fingerprint verification fails once.

Oh? I'll get on it.

**Comment by FabianHenneke** [comment]
> > I also found an issue we could and should fix ourselves: Our BiometricPrompt currently throws an exception when the fingerprint verification fails once.
> 
> Oh? I'll get on it.

I tried again and found that the following happens: If you touch the fingerprint sensor with e.g. a finger that hasn't been registered, it does not let you retry but fails. That failure is then not handled gracefully by this PR, which I'll take care off. But I think that we should have the prompt retry and not fail on the first attempt if it did acquire an image (which may however not be satisfactory).

**Comment by FabianHenneke** [comment]
> > I merged and confirmed that this works with RSA keys. Something in SSHJ seems to have broken ecdsa keys, I will look into that later, but you should be able to test with RSA keys.
> 
> Alright.

Turns out that ecdsa keys do work, it's just that SSHJ throws a really misleading exception if the remote doesn't accept the ecdsa public key.

**Comment by FabianHenneke** [comment]
> > > I also found an issue we could and should fix ourselves: Our BiometricPrompt currently throws an exception when the fingerprint verification fails once.
> > 
> > 
> > Oh? I'll get on it.
> 
> I tried again and found that the following happens: If you touch the fingerprint sensor with e.g. a finger that hasn't been registered, it does not let you retry but fails. That failure is then not handled gracefully by this PR, which I'll take care off. But I think that we should have the prompt retry and not fail on the first attempt if it did acquire an image (which may however not be satisfactory).

On a Pixel 3a running the Android 11 Beta, `onAuthenticationFailed` is called after a single fingerprint scan fails. On OnePlus running Android 10, the prompt allows retries and only shows a "Not recognized" method. Weird...

**Comment by FabianHenneke** [comment]
> > > I merged and confirmed that this works with RSA keys. Something in SSHJ seems to have broken ecdsa keys, I will look into that later, but you should be able to test with RSA keys.
> > 
> > 
> > Alright.
> 
> Turns out that ecdsa keys do work, it's just that SSHJ throws a really misleading exception if the remote doesn't accept the ecdsa public key.

Turns out that ecdsa keys **don't** work unless the server has an ecdsa host key. This is an upstream bug that got introduced while adding support for SHA-256/512 RSA and is now being tracked in https://github.com/hierynomus/sshj/issues/600.

**Comment by FabianHenneke** [comment]
It seems likely that this will not be ready in time for the June release. The upstream bug would need to be fixed first, including a release. 

I would prefer to update the FAQ for the next release. For now, nothing much has changed that is user-facing and I don't want to have users import their ECC keys now when the next release will feature secure on-device generation.

**Comment by msfjarvis** [comment]
> It seems likely that this will not be ready in time for the June release. The upstream bug would need to be fixed first, including a release.
> 
> I would prefer to update the FAQ for the next release. For now, nothing much has changed that is user-facing and I don't want to have users import their ECC keys now when the next release will feature secure on-device generation.

Yeah I agree. Let's draft this for now and get back once the upstream blockers are resolved.

**Comment by msfjarvis** [comment]
Not sure how well I did on the blind merge but this should be okay. I'm gonna email Jeroen and see if we can somehow expedite a 0.30.0 release.

**Comment by fmeum** [comment]
> Not sure how well I did on the blind merge but this should be okay. I'm gonna email Jeroen and see if we can somehow expedite a 0.30.0 release.

Unfortunately, a release at this point wouldn't help too much: The issues https://github.com/hierynomus/sshj/issues/600 and https://github.com/hierynomus/sshj/issues/608 are still open, only the former has a PR to fix it. These would break the new SHA256 and SHA512 signature types and possibly the EC algorithms for us.

**Comment by fmeum** [comment]
No longer blocked on https://github.com/hierynomus/sshj/issues/608 since we could work around this issue on our side.

**Comment by fmeum** [comment]
I changed the config to be compatible with the concept of key algorithms introduced in SSHJ 0.30.0.

**Comment by fmeum** [comment]
I will radically simplify the algorithm selection and remove Strongbox support since on second thought I fail to see how it has a strictly positive effect on any reasonable threat model. The new UI will briefly explain pros and cons of the only two choices (RSA 3072 and P-384).

**Comment by fmeum** [comment]
While working on v2 of this PR, I came up with an idea: We could offer ed25519 by wrapping the secret key with a Keystore-backed `EncryptedFile`, which is probably strictly better than having users generate the key on their computer and import it into APS. Should I add this?

**Comment by msfjarvis** [comment]
> While working on v2 of this PR, I came up with an idea: We could offer ed25519 by wrapping the secret key with a Keystore-backed `EncryptedFile`, which is probably strictly better than having users generate the key on their computer and import it into APS. Should I add this?

Seems like a good option, certainly in favor of adding.

**Comment by fmeum** [comment]
Superseded by #1070.

**Pull request closed without merging**
